dbms/src/Common/OptimizedRegularExpression.inl.h

```diff
@@ -725,18 +737,80 @@ void OptimizedRegularExpressionImpl<thread_safe>::replace(
     Int64 occur)
 {
     Int64 utf8_total_len = DB::UTF8::countCodePoints(reinterpret_cast<const UInt8 *>(subject), subject_size);
-    ;
+
     FunctionsRegexp::checkArgsReplace(utf8_total_len, subject_size, pos);
     FunctionsRegexp::makeReplaceOccurValid(occur);
 
+    auto num_captures = std::min(re2->NumberOfCapturingGroups() + 1, max_captures);
+    Instructions instructions = getInstructions(repl, num_captures);
+
     if (unlikely(subject_size == 0))
     {
-        processReplaceEmptyStringExpr(subject, subject_size, res_data, res_offset, repl, pos, occur);
+        processReplaceEmptyStringExpr(subject, subject_size, res_data, res_offset, pos, occur, num_captures, instructions);
         return;
     }
 
     size_t byte_pos = DB::UTF8::utf8Pos2bytePos(reinterpret_cast<const UInt8 *>(subject), pos);
-    replaceImpl(subject, subject_size, res_data, res_offset, repl, byte_pos, occur);
+    replaceImpl(subject, subject_size, res_data, res_offset, byte_pos, occur, num_captures, instructions);
+}
+
+template <bool thread_safe>
+Instructions OptimizedRegularExpressionImpl<thread_safe>::getInstructions(const StringRef & repl, int num_captures)
+{
+    Instructions instructions;
+    String literals;
+
+    for (size_t i = 0; i < repl.size; ++i)
+    {
+        if (repl.data[i] == '$' && i + 1 < repl.size)
+        {
+            if (isNumericASCII(repl.data[i + 1])) /// Substitution
+            {
+                if (!literals.empty())
+                {
+                    instructions.emplace_back(literals);
+                    literals = "";
+                }
+                instructions.emplace_back(repl.data[i + 1] - '0');
+            }
+            else
+                literals += repl.data[i + 1]; /// Escaping
```

Comment:
This implicitly assume that `$` is the escaping char, is it the expected behavior?

Reply:
> This implicitly assume that `$` is the escaping char, is it the expected behavior?

Yes, so does MySQL do.